src/lxc/lxc_attach.c

```diff
@@ -273,7 +273,12 @@ static int get_pty_on_host(struct lxc_container *c, struct wrapargs *wrap, int *
 	if (lxc_console_create(conf) < 0)
 		return -1;
 	ts = conf->console.tty_state;
-	ts->stdoutfd = STDOUT_FILENO;
+	/* If stdout is not a tty but stderr is, we write to stderr. */
+	if (!isatty(STDOUT_FILENO) && isatty(STDERR_FILENO))
+		ts->stdoutfd = STDERR_FILENO;
+	else
+		ts->stdoutfd = STDOUT_FILENO;
+
```

Comment:
Agreed it is confusing, and I'm only paying half attention so should probably let Christian explain.
The fundamental reason this had to change was that, for security reasons, we needed to interject a new pty and insulate lxc-attach's caller's tty from the container.
If any caller's fd 0-2 is not tty, then it is *not* redirected.  If any is, then a new pty is dup'ed over it.
So when you lxc-attach -n x1  ... < a > b, then stderr will point to a new pty, but 0 and 1 will point to a and b, respectively.  if you lxc-attach -n x1 ... < a, then 1 and 2 will both be dups of the new pty, and in that case stdout and stderr are combined.